[19.0][MIG] contract_variable_quantity

bosd opened the pull request:
Standard migration using odoo-module-migrator 
/ocabot migration contract_variable_quantity .
depends on #1312

pedrobaeza committed:
[ADD] contract_variable_quantity:

=================================================
Variable quantity in contract recurrent invoicing
=================================================

With this module, you will be able to define in recurring contracts some
lines with variable quantity according a provided formula.

Configuration
=============

* Go to Sales > Configuration > Contracts > Formulas (quantity).
* Define any formula based on Python code that stores at some moment a
  float/integer value of the quantity to invoice in the variable 'result'.

  You can use these variables to compute your formula:

  * *env*: Environment variable for getting other models.
  * *context*: Current context dictionary.
  * *user*: Current user.
  * *line*: Contract recurring invoice line that triggers this formula.
  * *contract*: Contract whose line belongs to.
  * *invoice*: Invoice (header) being created.

Usage
=====

To use this module, you need to:

* Go to Sales -> Contracts and select or create a new contract.
* Check *Generate recurring invoices automatically*.
* Add a new recurring invoicing line.
* Select "Variable quantity" in column "Qty. type".
* Select one of the possible formulas to use (previously created).

pedrobaeza committed:
[ADD] contract_variable_quantity:

=================================================
Variable quantity in contract recurrent invoicing
=================================================

With this module, you will be able to define in recurring contracts some
lines with variable quantity according a provided formula.

Configuration
=============

* Go to Sales > Configuration > Contracts > Formulas (quantity).
* Define any formula based on Python code that stores at some moment a
  float/integer value of the quantity to invoice in the variable 'result'.

  You can use these variables to compute your formula:

  * *env*: Environment variable for getting other models.
  * *context*: Current context dictionary.
  * *user*: Current user.
  * *line*: Contract recurring invoice line that triggers this formula.
  * *contract*: Contract whose line belongs to.
  * *invoice*: Invoice (header) being created.

Usage
=====

To use this module, you need to:

* Go to Sales -> Contracts and select or create a new contract.
* Check *Generate recurring invoices automatically*.
* Add a new recurring invoicing line.
* Select "Variable quantity" in column "Qty. type".
* Select one of the possible formulas to use (previously created).

pedrobaeza committed:
[IMP] contract_variable_quantity: Screenshot

pedrobaeza committed:
[IMP] contract_variable_quantity: Screenshot

pedrobaeza committed:
[IMP] contract_variable_quantity: Spanish translation

pedrobaeza committed:
[IMP] contract_variable_quantity: Spanish translation

carlosdauden committed:
[IMP] contract: Add past receipt type. Fix yearly. Add month last day

carlosdauden committed:
[IMP] contract: Add past receipt type. Fix yearly. Add month last day

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Rename manifest files

pedrobaeza committed:
[MIG] Rename manifest files

lasley committed:
[MIG] contract_variable_quantity: Migrate to v10
* Bump version
* Change openerp references to odoo
* Installable to True
* Change permissions & views to accounting dependency

lasley committed:
[MIG] contract_variable_quantity: Migrate to v10
* Bump version
* Change openerp references to odoo
* Installable to True
* Change permissions & views to accounting dependency

lasley committed:
[IMP][10.0] contract: Add templating (#42)

Add template functionality for contracts

lasley committed:
[IMP][10.0] contract: Add templating (#42)

Add template functionality for contracts

lasley committed:
[10.0][IMP] Add variable qty to contract template view (#66)

* [IMP] Add variable qty to contract template view

* Bump minor version

lasley committed:
[10.0][IMP] Add variable qty to contract template view (#66)

* [IMP] Add variable qty to contract template view

* Bump minor version

moylop260 committed:
[FIX] contract_payment_auto: Fix freze wkhtmltopdf freezed

moylop260 committed:
[FIX] contract_payment_auto: Fix freze wkhtmltopdf freezed

carlosdauden committed:
[MIG] contract_variable_quantity: Migration to 11.0

carlosdauden committed:
[MIG] contract_variable_quantity: Migration to 11.0

pedrobaeza committed:
[FIX] contract_variable_quantity: Correct README menus references

pedrobaeza committed:
[FIX] contract_variable_quantity: Correct README menus references

carlosdauden committed:
[IMP] contract_variable_quantity: Add option to don't create zero qty invoice lines

carlosdauden committed:
[IMP] contract_variable_quantity: Add option to don't create zero qty invoice lines

pedrobaeza committed:
[IMP] contract_variable_quantity: Make name translatable

pedrobaeza committed:
[IMP] contract_variable_quantity: Make name translatable

pedrobaeza committed:
[IMP] contract_variable_quantity: Recompute price when changing qty

If you have contract lines with automatic price and your pricelist
contains different prices per quantity, the price is not changed
when computing quantity.

This PR fixes this.

pedrobaeza committed:
[IMP] contract_variable_quantity: Recompute price when changing qty

If you have contract lines with automatic price and your pricelist
contains different prices per quantity, the price is not changed
when computing quantity.

This PR fixes this.

katyukha committed:
[FIX] analytic invoice/contract lines inheritance

Bug description
---------------

`account.analytic.contract.line` inherits
`account.analytic.invoice.line`

`account.analytic.invoice.line` defines field `analytic_account_id`:
   - comodel='account.analytic.account'

`account.analytic.contract.line` redefines field `analytic_account_id`:
   - comodel='account.analytic.contract'

On attempt to extend `account.analytic.invoice.line` model adding
field that depends on `analytic_account_id.date_start`
Odoo fails to update, because it adds this field to
`account.analytic.contract.line` through inheritance,
and `account.analytic.contract` model have no this field.

What is done
------------

Change inheritance order:
- `account.analytic.invoice.line` inherits
`account.analytic.contract.line`
- no file renames at this stage (this wil be done in next commit)

katyukha committed:
[FIX] analytic invoice/contract lines inheritance

Bug description
---------------

`account.analytic.contract.line` inherits
`account.analytic.invoice.line`

`account.analytic.invoice.line` defines field `analytic_account_id`:
   - comodel='account.analytic.account'

`account.analytic.contract.line` redefines field `analytic_account_id`:
   - comodel='account.analytic.contract'

On attempt to extend `account.analytic.invoice.line` model adding
field that depends on `analytic_account_id.date_start`
Odoo fails to update, because it adds this field to
`account.analytic.contract.line` through inheritance,
and `account.analytic.contract` model have no this field.

What is done
------------

Change inheritance order:
- `account.analytic.invoice.line` inherits
`account.analytic.contract.line`
- no file renames at this stage (this wil be done in next commit)

katyukha committed:
Update addon versions

- contract
- contract_variable_quantity

katyukha committed:
Update addon versions

- contract
- contract_variable_quantity

sbejaoui committed:
[MIG] contract_variable_quantity: Migration to 12.0

sbejaoui committed:
[MIG] contract_variable_quantity: Migration to 12.0

sbejaoui committed:
[IMP] Adapt unit test

sbejaoui committed:
[IMP] Adapt unit test

sbejaoui committed:
[IMP] - Add ace-widget with python mode

sbejaoui committed:
[IMP] - Add ace-widget with python mode

sbejaoui committed:
[IMP] - Adapt contract_variable_quantity to contract new views

sbejaoui committed:
[IMP] - Adapt contract_variable_quantity to contract new views

sbejaoui committed:
[IMP] - refresh contract_variable_quantity.pot

sbejaoui committed:
[IMP] - refresh contract_variable_quantity.pot

ThomasBinsfeld committed:
[REF] Contract: invoice in prepare_invoice_line is optional

ThomasBinsfeld committed:
[REF] Contract: invoice in prepare_invoice_line is optional

sbejaoui committed:
[IMP] - Skip zero should ignore lines with qty zero even for fixed qty

sbejaoui committed:
[IMP] - Skip zero should ignore lines with qty zero even for fixed qty

sbejaoui committed:
[FIX] - Fix unit tests

sbejaoui committed:
[FIX] - Fix unit tests

sbejaoui committed:
[FIX] fix unit tests

sbejaoui committed:
[FIX] fix unit tests

sbejaoui committed:
[FIX] - fix check quantity = 0

sbejaoui committed:
[FIX] - fix check quantity = 0

sbejaoui committed:
[REF] - Use method _get_quantity_to_invoice.

sbejaoui committed:
[REF] - Use method _get_quantity_to_invoice.

sbejaoui committed:
[12.0][REF] - Contract Variable quantity: split from analytic accoun

sbejaoui committed:
[12.0][REF] - Contract Variable quantity: split from analytic accoun

gurneyalex committed:
[FIX] add missing newlines in readme files

gurneyalex committed:
[FIX] add missing newlines in readme files

sbejaoui committed:
[FIX] - redundant fields in contract line form view

sbejaoui committed:
[FIX] - redundant fields in contract line form view

pedrobaeza committed:
[FIX] contract_variable_quantity: Allow to propagate variable qty fields from template

**Steps to reproduce the problem**

- Define a contract template with one line with variable quantity and formula.
- Use that template in a new contract.
- Variable quantity type and formula are not transferred from template to contract.

This is due to the lack of that fields in the tree view that prevents web client to fetch
and store them.

We solve it adding those fields to the contract line tree view, also providing some
visibility logic for not showing the corresponding fields according the type.

No regression tests is provided, as this is only reproducible with web client, not through
code.

pedrobaeza committed:
[FIX] contract_variable_quantity: Allow to propagate variable qty fields from template

**Steps to reproduce the problem**

- Define a contract template with one line with variable quantity and formula.
- Use that template in a new contract.
- Variable quantity type and formula are not transferred from template to contract.

This is due to the lack of that fields in the tree view that prevents web client to fetch
and store them.

We solve it adding those fields to the contract line tree view, also providing some
visibility logic for not showing the corresponding fields according the type.

No regression tests is provided, as this is only reproducible with web client, not through
code.

pedrobaeza committed:
[IMP] contract: remove redundant method

I have detected a method that was created as redundant and with the same
technique used when preparing the line values, so better to have
everything together in the same method instead of having it spread.

pedrobaeza committed:
[IMP] contract: remove redundant method

I have detected a method that was created as redundant and with the same
technique used when preparing the line values, so better to have
everything together in the same method instead of having it spread.

Commit:
[IMP] contract: add section and notes in contract line

Commit:
[IMP] contract: add section and notes in contract line

CarlosRoca13 committed:
[IMP] contract_variable_quantity: black, isort, prettier

CarlosRoca13 committed:
[IMP] contract_variable_quantity: black, isort, prettier

CarlosRoca13 committed:
[MIG] contract_variable_quantity: Migration to v13

CarlosRoca13 committed:
[MIG] contract_variable_quantity: Migration to v13

victoralmau committed:
[IMP] contract_variable_quantity: Add qty_type field in portal

victoralmau committed:
[IMP] contract_variable_quantity: Add qty_type field in portal

angelmoya committed:
FIX variable quantity view

angelmoya committed:
FIX variable quantity view

victoralmau committed:
[IMP] contract_variable_quantity: black, isort, prettier

victoralmau committed:
[IMP] contract_variable_quantity: black, isort, prettier

victoralmau committed:
[MIG] contract_variable_quantity: Migration to 14.0

[UPD] Update contract_variable_quantity.pot

[UPD] README.rst

victoralmau committed:
[MIG] contract_variable_quantity: Migration to 14.0

[UPD] Update contract_variable_quantity.pot

[UPD] README.rst

victoralmau committed:
[MIG] contract_variable_quantity: Migration to 15.0

TT40289

victoralmau committed:
[MIG] contract_variable_quantity: Migration to 15.0

TT40289

carolinafernandez-tecnativa committed:
[MIG] contract_variable_quantity: Migration to 16.0

carolinafernandez-tecnativa committed:
[MIG] contract_variable_quantity: Migration to 16.0

aliciagaarzo committed:
[16.0][IMP] contract and contract_variable_quantity: improve visual of contract lines

aliciagaarzo committed:
[16.0][IMP] contract and contract_variable_quantity: improve visual of contract lines

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (35 of 35 strings)

Translation: contract-16.0/contract-16.0-contract_variable_quantity
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_variable_quantity/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (35 of 35 strings)

Translation: contract-16.0/contract-16.0-contract_variable_quantity
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_variable_quantity/it/

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 68.5% (24 of 35 strings)

Translation: contract-16.0/contract-16.0-contract_variable_quantity
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_variable_quantity/hr/

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 68.5% (24 of 35 strings)

Translation: contract-16.0/contract-16.0-contract_variable_quantity
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_variable_quantity/hr/

jakobkrabbe committed:
Translated using Weblate (Swedish)

Currently translated at 100.0% (35 of 35 strings)

Translation: contract-16.0/contract-16.0-contract_variable_quantity
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_variable_quantity/sv/

jakobkrabbe committed:
Translated using Weblate (Swedish)

Currently translated at 100.0% (35 of 35 strings)

Translation: contract-16.0/contract-16.0-contract_variable_quantity
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_variable_quantity/sv/

juanjosesegui-tecnativa committed:
[IMP] contract_variable_quantity: pre-commit auto fixes

juanjosesegui-tecnativa committed:
[IMP] contract_variable_quantity: pre-commit auto fixes

juanjosesegui-tecnativa committed:
[MIG] contract_variable_quantity: Migration to 17.0

juanjosesegui-tecnativa committed:
[MIG] contract_variable_quantity: Migration to 17.0

sbejaoui committed:
[18.0][MIG] contract_variable_quantity

sbejaoui committed:
[18.0][MIG] contract_variable_quantity

bo-eska committed:
Translated using Weblate (Turkish)

Currently translated at 97.0% (33 of 34 strings)

Translation: contract-18.0/contract-18.0-contract_variable_quantity
Translate-URL: https://translation.odoo-community.org/projects/contract-18-0/contract-18-0-contract_variable_quantity/tr/

bo-eska committed:
Translated using Weblate (Turkish)

Currently translated at 97.0% (33 of 34 strings)

Translation: contract-18.0/contract-18.0-contract_variable_quantity
Translate-URL: https://translation.odoo-community.org/projects/contract-18-0/contract-18-0-contract_variable_quantity/tr/

bosd committed:
[MIG] contract_variable_quantity: Migration to 19.0

bosd committed:
[MIG] contract_variable_quantity: Migration to 19.0

bosd committed:
[TEST REQUIREMENTS DO NOT MERGE]

bosd commented:
/ocabot migration contract_variable_quantity

yvaucher commented:
/ocabot migration contract_variable_quantity

carlos-lopez-tecnativa commented on contract_variable_quantity/models/contract_line.py:
```diff
@@ -0,0 +1,54 @@
+        return quantity
+
+    def _prepare_invoice_line(self):
+        vals = super()._prepare_invoice_line()
+        if (
+            "quantity" in vals
+            and self.contract_id.skip_zero_qty
+            and float_is_zero(
+                vals["quantity"],
+                self.env["decimal.precision"].precision_get("Product Unit of Measure"),
+            )
+        ):
+            vals = {}
```
Please take sections and notes into account. These lines should also be included, even though their quantity is zero.

When I enable this option, those lines are currently not included in the invoice.

**Contract**
<img width="1241" height="805" alt="Image" src="https://github.com/user-attachments/assets/6b660216-37cb-4023-9bc7-ea2ac58ccd86" />

**Invoice**
<img width="1547" height="833" alt="Image" src="https://github.com/user-attachments/assets/d29db397-8aef-4776-969e-6b038ef71956" />

bosd replied:
Good catch — fixed. The `skip_zero_qty` guard now excludes `display_type` lines, so sections and notes (which have no quantity) are always kept on the invoice. Added a test (`test_skip_zero_qty_keeps_sections_and_notes`) covering it.

carlos-lopez-tecnativa reviewed (changes requested):
Please rebase the branch and remove the last commit, as the dependency has already been merged.

TT58613

Commit:
[FIX] contract_variable_quantity: keep sections and notes when skipping zero-qty lines

The skip_zero_qty option blanks any invoice line whose quantity is zero.
Section and note lines (display_type set) also have a zero quantity, so
they were dropped from the invoice too. Guard the skip with
`not self.display_type` so structural lines are always kept.

carlos-lopez-tecnativa reviewed (approved):
LGTM
@pedrobaeza @victoralmau could you please review this?

pedrobaeza reviewed (approved):
/ocabot merge nobump
